File: cpp/이분탐색_LIS/기타레슨.cpp
#include <iostream>
#include <vector>
typedef long long ll;
using namespace std;

ll n, m, ret = -1, mx = -1;
vector<ll> v;

bool check(ll mid){
	if(mx > mid) return false;

	int temp = mid;
	int cnt = 0;
	for(int i = 0; i < n; i++){
		if(mid - v[i] < 0){
			mid = temp;
			cnt++;
		}
		mid -= v[i];
	}
	if(mid != temp) cnt++;
	return cnt <= m;
}
int main(){
	cin >> n >> m;
	ll tmp;
	ll lo = 0, hi = 0;
	for(int i = 0 ; i < n; i++) {
		cin >> tmp; v.push_back(tmp);
		hi += v[i];
		mx = max(mx, tmp);
	}

	while(lo <= hi){
		ll mid = (lo + hi) / 2;
		if(check(mid)){
			hi = mid - 1;
			ret = mid;
		}else{
			lo = mid + 1;
		}
	}

	cout << ret << '\n';
}

File: python/백준/스터디/20회/기타레슨.py
import sys
input = sys.stdin.readline

n,m = map(int,input().split())
arr = list(map(int,input().split()))
l,r,s,mx = 1,sum(arr),sum(arr), max(arr)
ret = sys.maxsize

def check(mid):
    if mid < mx:
        return False
    cnt = 0
    sum = 0
    # print('mid==========',mid)
    for i in arr:
        sum += i
        if sum > mid:
            cnt += 1
            sum = i
        elif sum == mid:
            cnt += 1
            sum = 0
        # print(sum)
    if sum > 0:
        cnt +=1
    # print('cnt = ',cnt)
    if cnt <= m:
        return True
    return False

while l <= r:
    mid = (l+r) // 2
    if check(mid):
        # print('가능!', mid)
        r = mid - 1
        ret = mid
    else:
        # print('불가능!', mid)
        l = mid + 1
print(ret)